File: sdk/typescript/tests/workspace.test.ts
/**
 * SPDX-License-Identifier: Apache-2.0
 * Copyright 2026 The Plinth Authors
 */

import { describe, expect, it } from "vitest";

import {
  KeyNotFoundError,
  Plinth,
  type FileEntry,
  type KVEntry,
  type Snapshot,
  type Workspace,
} from "../src/index.js";
import { MockServer, parseQuery } from "./_helpers.js";

async function bootstrap(server: MockServer): Promise<{ client: Plinth; ws: Workspace }> {
  server.json("GET", /\/v1\/workspaces$/, {
    workspaces: [
      {
        id: "ws_1",
        name: "alpha",
        created_at: "2026-01-01T00:00:00Z",
        updated_at: "2026-01-01T00:00:00Z",
        metadata: {},
      },
    ],
  });
  const client = new Plinth({
    workspaceUrl: "http://workspace.test",
    gatewayUrl: "http://gateway.test",
    apiKey: "test-token",
    fetch: server.fetch as unknown as typeof fetch,
  });
  const ws = await client.workspace("alpha");
  return { client, ws };
}

describe("Workspace KV", () => {
  it("PUTs new versions and reads them back", async () => {
    const server = new MockServer();
    server.on("PUT", /\/kv\/topic$/, (req) => {
      const body = JSON.parse(req.body ?? "{}");
      const entry: KVEntry = {
        workspace_id: "ws_1",
        key: "topic",
        value: body.value,
        version: 1,
        created_at: "2026-01-01T00:00:00Z",
        deleted: false,
        branch_id: null,
      };
      return { body: entry };
    });
    server.on("GET", /\/kv\/topic(\?.*)?$/, () => ({
      body: {
        workspace_id: "ws_1",
        key: "topic",
        value: "renewable energy",
        version: 1,
        created_at: "2026-01-01T00:00:00Z",
        deleted: false,
        branch_id: null,
      },
    }));

    const { ws } = await bootstrap(server);
    const written = await ws.kv.set("topic", "renewable energy");
    expect(written.version).toBe(1);
    expect(written.value).toBe("renewable energy");

    const value = await ws.kv.get("topic");
    expect(value).toBe("renewable energy");

    const meta = await ws.kv.getWithMeta("topic");
    expect(meta.version).toBe(1);
  });

  it("returns null on a tombstoned key", async () => {
    const server = new MockServer();
    server.on("GET", /\/kv\/dead/, () => ({
      body: {
        workspace_id: "ws_1",
        key: "dead",
        value: null,
        version: 2,
        created_at: "2026-01-01T00:00:00Z",
        deleted: true,
        branch_id: null,
      },
    }));

    const { ws } = await bootstrap(server);
    expect(await ws.kv.get("dead")).toBeNull();
  });

  it("history returns all versions", async () => {
    const server = new MockServer();
    server.json("GET", /\/kv\/topic\/history/, {
      versions: [
        { workspace_id: "ws_1", key: "topic", value: "a", version: 1, created_at: "2026-01-01T00:00:00Z", deleted: false, branch_id: null },
        { workspace_id: "ws_1", key: "topic", value: "b", version: 2, created_at: "2026-01-01T00:00:01Z", deleted: false, branch_id: null },
      ],
    });
    const { ws } = await bootstrap(server);
    const history = await ws.kv.history("topic");
    expect(history).toHaveLength(2);
    expect(history[1]!.value).toBe("b");
  });

  it("propagates KeyNotFoundError", async () => {
    const server = new MockServer();
    server.on("GET", /\/kv\/missing/, () => ({
      status: 404,
      body: { error: { code: "KEY_NOT_FOUND", message: "no such key" } },
    }));
    const { ws } = await bootstrap(server);
    await expect(ws.kv.get("missing")).rejects.toBeInstanceOf(KeyNotFoundError);
  });

  it("DELETE issues a tombstone", async () => {
    const server = new MockServer();
    server.on("DELETE", /\/kv\/dead/, () => ({ status: 204 }));
    const { ws } = await bootstrap(server);
    await expect(ws.kv.delete("dead")).resolves.toBeUndefined();
    expect(server.requests.find((r) => r.method === "DELETE")).toBeDefined();
  });

  it("encodes keys with special characters", async () => {
    const server = new MockServer();
    server.json("PUT", /\/kv\/key%20with%20space/, {
      workspace_id: "ws_1",
      key: "key with space",
      value: 1,
      version: 1,
      created_at: "2026-01-01T00:00:00Z",
      deleted: false,
      branch_id: null,
    });
    const { ws } = await bootstrap(server);
    await ws.kv.set("key with space", 1);
    expect(server.requests.some((r) => r.url.includes("key%20with%20space"))).toBe(true);
  });
});

describe("Workspace Files", () => {
  it("writes string content as bytes and reads back text", async () => {
    const server = new MockServer();
    server.on("PUT", /\/files\/report\.md/, () => ({
      body: {
        workspace_id: "ws_1",
        path: "report.md",
        size: 13,
        sha256: "deadbeef",
        content_type: "text/plain; charset=utf-8",
        version: 1,
        created_at: "2026-01-01T00:00:00Z",
        deleted: false,
        branch_id: null,
      } as FileEntry,
    }));
    server.on("GET", /\/files\/report\.md/, (req) => {
      // Don't match /meta route
      if (req.url.endsWith("/meta")) return { status: 404, body: { error: { code: "FILE_NOT_FOUND", message: "" } } };
      return { bodyBytes: new TextEncoder().encode("# Report\n...") };
    });

    const { ws } = await bootstrap(server);
    const meta = await ws.files.write("report.md", "# Report\n...");
    expect(meta.size).toBe(13);

    const text = await ws.files.readText("report.md");
    expect(text).toBe("# Report\n...");
  });

  it("preserves nested file paths", async () => {
    const server = new MockServer();
    server.on("PUT", /\/files\/notes\/2026\/q1\.md/, () => ({
      body: {
        workspace_id: "ws_1",
        path: "notes/2026/q1.md",
        size: 5,
        sha256: "x",
        content_type: "text/plain; charset=utf-8",
        version: 1,
        created_at: "2026-01-01T00:00:00Z",
        deleted: false,
        branch_id: null,
      } as FileEntry,
    }));
    const { ws } = await bootstrap(server);
    await ws.files.write("notes/2026/q1.md", "hello");
    const matched = server.requests.find((r) => r.url.includes("/files/notes/2026/q1.md"));
    expect(matched).toBeDefined();
  });
});

describe("Workspace snapshots and branches", () => {
  it("creates a snapshot with optional message", async () => {
    const server = new MockServer();
    server.on("POST", /\/snapshots$/, (req) => {
      const body = JSON.parse(req.body ?? "{}");
      const snap: Snapshot = {
        id: "snap_1",
        workspace_id: "ws_1",
        name: body.name,
        message: body.message ?? null,
        created_at: "2026-01-01T00:00:00Z",
        kv_versions: {},
        file_versions: {},
        parent_snapshot_id: null,
      };
      return { status: 201, body: snap };
    });
    const { ws } = await bootstrap(server);
    const snap = await ws.snapshot("baseline", { message: "initial state" });
    expect(snap.id).toBe("snap_1");
    expect(snap.message).toBe("initial state");
  });

  it("withBranch attaches ?branch= to KV reads/writes", async () => {
    const server = new MockServer();
    server.on("PUT", /\/kv\/topic/, (req) => {
      const q = parseQuery(req.url);
      expect(q.branch).toBe("br_1");
      return {
        body: {
          workspace_id: "ws_1",
          key: "topic",
          value: JSON.parse(req.body ?? "{}").value,
          version: 2,
          created_at: "2026-01-01T00:00:00Z",
          deleted: false,
          branch_id: "br_1",
        } as KVEntry,
      };
    });

    const { ws } = await bootstrap(server);
    const wsB = ws.withBranch("br_1");
    const written = await wsB.kv.set("topic", "branch value");
    expect(written.branch_id).toBe("br_1");
    // The original ws should still be branch-less
    expect(ws.branchId).toBeNull();
  });

  it("merge POSTs to the branch merge endpoint", async () => {
    const server = new MockServer();
    server.on("POST", /\/branches\/br_1\/merge/, () => ({
      body: { branch_id: "br_1", merged: true, merged_at: "2026-01-01T00:00:01Z", conflicts: [] },
    }));
    const { ws } = await bootstrap(server);
    const result = await ws.merge("br_1");
    expect(result.merged).toBe(true);
  });
});

File: sdk/go/plinth_test/workspace_test.go
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 The Plinth Authors

package plinth_test

import (
	"context"
	"errors"
	"strings"
	"testing"
	"time"

	"github.com/plinth/sdk-go/plinth"
)

// workspaceFixture wires a MockServer with the standard
// "GET /v1/workspaces returns one workspace" + "POST /v1/workspaces
// creates" routes so individual tests can call client.Workspace(...)
// without re-stating the get-or-create dance.
//
// Returns the *Plinth, the *WorkspaceClient, and the MockServer for
// further route registration.
func workspaceFixture(t *testing.T) (*plinth.Plinth, *plinth.WorkspaceClient, *MockServer) {
	t.Helper()
	ms := NewMockServer(t)
	now := time.Now().Format(time.RFC3339Nano)
	ms.JSON("GET", "/v1/workspaces", 200, map[string]any{
		"workspaces": []any{
			map[string]any{
				"id":         "ws_test",
				"name":       "test-ws",
				"created_at": now,
				"updated_at": now,
			},
		},
	})

	c := newTestClient(t, ms)
	ws, err := c.Workspace(context.Background(), "test-ws")
	if err != nil {
		t.Fatalf("Workspace: %v", err)
	}
	return c, ws, ms
}

// TestKVSetAndGet covers the basic write/read round-trip plus version
// disclosure via GetWithVersion.
func TestKVSetAndGet(t *testing.T) {
	_, ws, ms := workspaceFixture(t)
	now := time.Now().Format(time.RFC3339Nano)

	ms.JSON("PUT", "/v1/workspaces/ws_test/kv/topic", 200, map[string]any{
		"workspace_id": "ws_test",
		"key":          "topic",
		"value":        "renewable energy",
		"version":      1,
		"created_at":   now,
		"deleted":      false,
	})
	ms.JSON("GET", "/v1/workspaces/ws_test/kv/topic", 200, map[string]any{
		"workspace_id": "ws_test",
		"key":          "topic",
		"value":        "renewable energy",
		"version":      1,
		"created_at":   now,
		"deleted":      false,
	})

	if _, err := ws.KV.Set(context.Background(), "topic", "renewable energy"); err != nil {
		t.Fatalf("KV.Set: %v", err)
	}
	val, err := ws.KV.Get(context.Background(), "topic")
	if err != nil {
		t.Fatalf("KV.Get: %v", err)
	}
	if val != "renewable energy" {
		t.Errorf("KV.Get = %v, want \"renewable energy\"", val)
	}

	val2, version, err := ws.KV.GetWithVersion(context.Background(), "topic")
	if err != nil {
		t.Fatalf("KV.GetWithVersion: %v", err)
	}
	if val2 != "renewable energy" || version != 1 {
		t.Errorf("GetWithVersion = (%v, %d), want (renewable energy, 1)", val2, version)
	}
}

// TestKVHistory covers the /history endpoint.
func TestKVHistory(t *testing.T) {
	_, ws, ms := workspaceFixture(t)
	now := time.Now().Format(time.RFC3339Nano)
	ms.JSON("GET", "/v1/workspaces/ws_test/kv/topic/history", 200, map[string]any{
		"versions": []any{
			map[string]any{"workspace_id": "ws_test", "key": "topic", "value": "v1", "version": 1, "created_at": now, "deleted": false},
			map[string]any{"workspace_id": "ws_test", "key": "topic", "value": "v2", "version": 2, "created_at": now, "deleted": false},
		},
	})
	versions, err := ws.KV.History(context.Background(), "topic")
	if err != nil {
		t.Fatalf("KV.History: %v", err)
	}
	if len(versions) != 2 {
		t.Fatalf("len(versions) = %d, want 2", len(versions))
	}
	if versions[1].Version != 2 {
		t.Errorf("versions[1].Version = %d, want 2", versions[1].Version)
	}
}

// TestKVGetNotFound exercises the 404 → typed-error path.
func TestKVGetNotFound(t *testing.T) {
	_, ws, ms := workspaceFixture(t)
	ms.Error("GET", "/v1/workspaces/ws_test/kv/missing", 404, "KEY_NOT_FOUND", "key missing not found")

	_, err := ws.KV.Get(context.Background(), "missing")
	if err == nil {
		t.Fatal("expected error, got nil")
	}
	if !errors.Is(err, plinth.ErrKeyNotFound) {
		t.Errorf("err = %v, want ErrKeyNotFound", err)
	}
}

// TestKVDelete covers the tombstone path.
func TestKVDelete(t *testing.T) {
	_, ws, ms := workspaceFixture(t)
	ms.On("DELETE", "/v1/workspaces/ws_test/kv/foo", func(_ recordedRequest) mockResponse {
		return mockResponse{Status: 204}
	})
	if err := ws.KV.Delete(context.Background(), "foo"); err != nil {
		t.Fatalf("KV.Delete: %v", err)
	}
}

// TestFilesWriteAndRead covers binary upload + download round-trip.
func TestFilesWriteAndRead(t *testing.T) {
	_, ws, ms := workspaceFixture(t)
	now := time.Now().Format(time.RFC3339Nano)

	ms.JSON("PUT", "/v1/workspaces/ws_test/files/report.md", 200, map[string]any{
		"workspace_id": "ws_test",
		"path":         "report.md",
		"size":         12,
		"sha256":       "deadbeef",
		"content_type": "text/plain; charset=utf-8",
		"version":      1,
		"created_at":   now,
		"deleted":      false,
	})
	ms.Bytes("GET", "/v1/workspaces/ws_test/files/report.md", 200, []byte("# Report\n..."), "text/plain; charset=utf-8")

	entry, err := ws.Files.WriteText(context.Background(), "report.md", "# Report\n...", nil)
	if err != nil {
		t.Fatalf("Files.WriteText: %v", err)
	}
	if entry.Version != 1 {
		t.Errorf("entry.Version = %d, want 1", entry.Version)
	}

	body, err := ws.Files.Read(context.Background(), "report.md")
	if err != nil {
		t.Fatalf("Files.Read: %v", err)
	}
	if string(body) != "# Report\n..." {
		t.Errorf("body = %q, want # Report\\n...", string(body))
	}

	// Verify Content-Type header on the upload.
	put := findRequest(ms.Requests(), "PUT", "/v1/workspaces/ws_test/files/report.md")
	if put == nil {
		t.Fatal("expected PUT request, got none")
	}
	if !strings.HasPrefix(put.Headers.Get("Content-Type"), "text/plain") {
		t.Errorf("Content-Type = %q, want text/plain prefix", put.Headers.Get("Content-Type"))
	}
}

// TestFilesReadText is the convenience wrapper for UTF-8 file reads.
func TestFilesReadText(t *testing.T) {
	_, ws, ms := workspaceFixture(t)
	ms.Bytes("GET", "/v1/workspaces/ws_test/files/notes.txt", 200, []byte("hello world"), "text/plain")

	got, err := ws.Files.ReadText(context.Background(), "notes.txt")
	if err != nil {
		t.Fatalf("Files.ReadText: %v", err)
	}
	if got != "hello world" {
		t.Errorf("ReadText = %q, want hello world", got)
	}
}

// TestFilesReadNotFound verifies the typed error when fetching a
// missing path.
func TestFilesReadNotFound(t *testing.T) {
	_, ws, ms := workspaceFixture(t)
	ms.Error("GET", "/v1/workspaces/ws_test/files/missing.txt", 404, "FILE_NOT_FOUND", "missing.txt")

	_, err := ws.Files.Read(context.Background(), "missing.txt")
	if !errors.Is(err, plinth.ErrFileNotFound) {
		t.Errorf("err = %v, want ErrFileNotFound", err)
	}
}

// TestSnapshotCreate covers the snapshot lifecycle.
func TestSnapshotCreate(t *testing.T) {
	_, ws, ms := workspaceFixture(t)
	now := time.Now().Format(time.RFC3339Nano)

	ms.JSON("POST", "/v1/workspaces/ws_test/snapshots", 201, map[string]any{
		"id":            "snap_1",
		"workspace_id":  "ws_test",
		"name":          "baseline",
		"message":       "initial state",
		"created_at":    now,
		"kv_versions":   map[string]any{},
		"file_versions": map[string]any{},
	})

	snap, err := ws.Snapshot(context.Background(), "baseline", "initial state")
	if err != nil {
		t.Fatalf("Snapshot: %v", err)
	}
	if snap.ID != "snap_1" {
		t.Errorf("snap.ID = %q, want snap_1", snap.ID)
	}

	post := findRequest(ms.Requests(), "POST", "/v1/workspaces/ws_test/snapshots")
	if post == nil {
		t.Fatal("expected POST snapshot request, got none")
	}
	var body map[string]any
	post.jsonBody(t, &body)
	if body["name"] != "baseline" || body["message"] != "initial state" {
		t.Errorf("body = %v, want name=baseline message=\"initial state\"", body)
	}
}

// TestBranchCreateAndList covers the branch endpoint pair.
func TestBranchCreateAndList(t *testing.T) {
	_, ws, ms := workspaceFixture(t)
	now := time.Now().Format(time.RFC3339Nano)

	ms.JSON("POST", "/v1/workspaces/ws_test/branches", 201, map[string]any{
		"id":               "br_1",
		"workspace_id":     "ws_test",
		"name":             "experiment",
		"from_snapshot_id": "snap_1",
		"created_at":       now,
		"merged":           false,
	})
	ms.JSON("GET", "/v1/workspaces/ws_test/branches", 200, map[string]any{
		"branches": []any{
			map[string]any{
				"id":               "br_1",
				"workspace_id":     "ws_test",
				"name":             "experiment",
				"from_snapshot_id": "snap_1",
				"created_at":       now,
				"merged":           false,
			},
		},
	})

	br, err := ws.Branch(context.Background(), "experiment", "snap_1")
	if err != nil {
		t.Fatalf("Branch: %v", err)
	}
	if br.ID != "br_1" {
		t.Errorf("br.ID = %q, want br_1", br.ID)
	}

	all, err := ws.Branches(context.Background())
	if err != nil {
		t.Fatalf("Branches: %v", err)
	}
	if len(all) != 1 || all[0].ID != "br_1" {
		t.Errorf("Branches = %v, want one branch br_1", all)
	}
}

// TestWithBranchScopesQuery verifies the WithBranch view appends
// ?branch=<id> to subsequent KV writes.
func TestWithBranchScopesQuery(t *testing.T) {
	_, ws, ms := workspaceFixture(t)
	now := time.Now().Format(time.RFC3339Nano)
	ms.JSON("PUT", "/v1/workspaces/ws_test/kv/foo", 200, map[string]any{
		"workspace_id": "ws_test", "key": "foo", "value": "v",
		"version": 1, "created_at": now, "deleted": false,
	})

	branched := ws.WithBranch("br_test")
	if branched.BranchID() != "br_test" {
		t.Errorf("BranchID = %q, want br_test", branched.BranchID())
	}
	if _, err := branched.KV.Set(context.Background(), "foo", "v"); err != nil {
		t.Fatalf("KV.Set on branch: %v", err)
	}
	last := ms.LastRequest(t)
	if !strings.Contains(last.Query, "branch=br_test") {
		t.Errorf("query = %q, want branch=br_test", last.Query)
	}
}

// TestChannelsSendAndReceive covers the v0.2 channel surface.
func TestChannelsSendAndReceive(t *testing.T) {
	_, ws, ms := workspaceFixture(t)
	now := time.Now().Format(time.RFC3339Nano)

	ms.JSON("POST", "/v1/workspaces/ws_test/channels/results/send", 201, map[string]any{
		"id":           "msg_1",
		"channel":      "results",
		"workspace_id": "ws_test",
		"seq":          1,
		"payload":      map[string]any{"text": "done"},
		"sent_at":      now,
	})
	ms.JSON("GET", "/v1/workspaces/ws_test/channels/results/receive", 200, map[string]any{
		"messages": []any{
			map[string]any{
				"id":           "msg_1",
				"channel":      "results",
				"workspace_id": "ws_test",
				"seq":          1,
				"payload":      map[string]any{"text": "done"},
				"sent_at":      now,
			},
		},
	})

	msg, err := ws.Channels.Send(context.Background(), "results",
		map[string]any{"text": "done"},
		plinth.ChannelSendOpts{Sender: "agent-A", Type: "result"},
	)
	if err != nil {
		t.Fatalf("Channels.Send: %v", err)
	}
	if msg.ID != "msg_1" {
		t.Errorf("msg.ID = %q, want msg_1", msg.ID)
	}

	got, err := ws.Channels.Receive(context.Background(), "results",
		plinth.ChannelReceiveOpts{Consumer: "writer", Limit: 10},
	)
	if err != nil {
		t.Fatalf("Channels.Receive: %v", err)
	}
	if len(got) != 1 || got[0].ID != "msg_1" {
		t.Errorf("got = %v, want one message msg_1", got)
	}

	// Verify ?consumer=writer&limit=10 was attached.
	rec := findRequest(ms.Requests(), "GET", "/v1/workspaces/ws_test/channels/results/receive")
	if rec == nil {
		t.Fatal("no receive request recorded")
	}
	if !strings.Contains(rec.Query, "consumer=writer") || !strings.Contains(rec.Query, "limit=10") {
		t.Errorf("query = %q, want consumer=writer&limit=10", rec.Query)
	}
}

// TestWorkflowsCreateAndStartStep covers a basic workflow round-trip.
func TestWorkflowsCreateAndStartStep(t *testing.T) {
	_, ws, ms := workspaceFixture(t)
	now := time.Now().Format(time.RFC3339Nano)

	ms.JSON("POST", "/v1/workspaces/ws_test/workflows", 201, map[string]any{
		"id":             "wf_1",
		"workspace_id":   "ws_test",
		"name":           "pipeline",
		"steps_manifest": []any{"a", "b", "c"},
		"steps":          []any{},
		"status":         "pending",
		"created_at":     now,
	})
	ms.JSON("POST", "/v1/workspaces/ws_test/workflows/wf_1/steps", 201, map[string]any{
		"id":          "step_a",
		"workflow_id": "wf_1",
		"name":        "a",
		"status":      "running",
		"attempt":     1,
		"started_at":  now,
	})

	wf, err := ws.Workflows.Create(context.Background(), "pipeline", []string{"a", "b", "c"})
	if err != nil {
		t.Fatalf("Workflows.Create: %v", err)
	}
	if wf.ID() != "wf_1" {
		t.Errorf("wf.ID = %q, want wf_1", wf.ID())
	}
	step, err := wf.StartStep(context.Background(), "a", map[string]any{"x": 1})
	if err != nil {
		t.Fatalf("StartStep: %v", err)
	}
	if step.ID != "step_a" || step.Status != plinth.WorkflowStatusRunning {
		t.Errorf("step = %+v, want id=step_a status=running", step)
	}
}

// TestWorkflowStartStepRejectsOffManifest exercises the client-side
// validation: an unknown step name fails before any HTTP call.
func TestWorkflowStartStepRejectsOffManifest(t *testing.T) {
	_, ws, ms := workspaceFixture(t)
	now := time.Now().Format(time.RFC3339Nano)
	ms.JSON("POST", "/v1/workspaces/ws_test/workflows", 201, map[string]any{
		"id":             "wf_1",
		"workspace_id":   "ws_test",
		"name":           "pipeline",
		"steps_manifest": []any{"a", "b"},
		"steps":          []any{},
		"status":         "pending",
		"created_at":     now,
	})

	wf, err := ws.Workflows.Create(context.Background(), "pipeline", []string{"a", "b"})
	if err != nil {
		t.Fatalf("Create: %v", err)
	}
	_, err = wf.StartStep(context.Background(), "z", nil)
	if err == nil {
		t.Fatal("expected error for off-manifest step name, got nil")
	}
	if !errors.Is(err, plinth.ErrInvalidWorkflowStep) {
		t.Errorf("err = %v, want ErrInvalidWorkflowStep", err)
	}
}

// TestWorkflowLeaseStepReturnsNilOnConflict checks that a 409
// LEASE_CONFLICT surfaces as (nil, nil) so worker code can branch
// cleanly without a typed-error import.
func TestWorkflowLeaseStepReturnsNilOnConflict(t *testing.T) {
	_, ws, ms := workspaceFixture(t)
	now := time.Now().Format(time.RFC3339Nano)
	ms.JSON("POST", "/v1/workspaces/ws_test/workflows", 201, map[string]any{
		"id":             "wf_1",
		"workspace_id":   "ws_test",
		"name":           "p",
		"steps_manifest": []any{"a"},
		"steps":          []any{},
		"status":         "pending",
		"created_at":     now,
	})
	ms.Error("POST", "/v1/workspaces/ws_test/workflows/wf_1/steps/step_a/lease",
		409, "LEASE_CONFLICT", "step already leased by another worker")

	wf, _ := ws.Workflows.Create(context.Background(), "p", []string{"a"})
	lease, err := wf.LeaseStep(context.Background(), "step_a", "worker-1", 60)
	if err != nil {
		t.Fatalf("LeaseStep returned err = %v on 409, want nil", err)
	}
	if lease != nil {
		t.Errorf("lease = %+v, want nil on 409", lease)
	}
}
